src/ledger.js

```diff
@@ -0,0 +1,131 @@
+                    }
+                })
+            }
+        })
+    }
+    getOperation(id) {
+        return this.db.get(Entities.OPERATIONS, id);
+    }
+    async postOperationEntries(operationId, entries) {
+        await sleep(2000);   // To test the background queue
+        this.db.update(Entities.OPERATIONS, operationId, {status: OperationStatus.PROCESSING})
+        // validate the entries using balance restrictions
+        try {
+            this.validateEntries(entries);
+        } catch (error) {
+            this.db.update(Entities.OPERATIONS, operationId, {status: OperationStatus.REJECTED, rejectionReason: error.message});
+            return;
+        }
+        this.db.insertMany(Entities.ENTRIES, entries);
+        this.db.update(Entities.OPERATIONS, operationId, {status: OperationStatus.APPLIED})
+    }
+    postOperation(operation) {
+        operation.status = OperationStatus.INIT;
+        const operationId = this.db.insert(Entities.OPERATIONS, operation);
+        this.postingQueue.enqueueTask(async () => {
+            await this.postOperationEntries(operationId, operation.entries);
+        });
+        return operationId;
+    }
+    async postOperationAsync(operation) {
+        return new Promise((resolve, reject) => {
+            operation.status = OperationStatus.INIT;
+            const operationId = this.db.insert(Entities.OPERATIONS, operation);
+            this.postingQueue.enqueueTask(async () => {
+                await this.postOperationEntries(operationId, operation.entries);
+                resolve(operationId);
+            });
+        })
+    }
+    postTransferOperation(transferData, async) {
+        const operation = {
+            operationType: OperationTypes.TRANSFER,
+            memo: transferData.memo,
+            entries: [
+                {
+                    accountId: transferData.fromAccountId,
+                    assetId: transferData.assetId,
+                    value: (-BigNumber(transferData.value)).toString()
+                },
+                {
+                    accountId: transferData.toAccountId,
+                    assetId: transferData.assetId,
+                    value: transferData.value
+                }
+            ]
+        }
+        if (async) {
+            return this.postOperationAsync(operation);
+        } else {
+            return this.postOperation(operation);
+        }
+    }
+    createAccount(accountInfo) {
+        return this.db.insert(Entities.ACCOUNTS, accountInfo);
+    }
+    getAccount(id) {
+        const account = this.db.get(Entities.ACCOUNTS, id);
+        if (account) {
+            account.balances = this.getAccountBalances(id);
+        }
+        return account;
+    }
+    getAccountBalances(id) {
+        const accountEntries = this.db.getAll(Entities.ENTRIES, id);
```

Comment:
one interesting complexity we have to model - 

we're saying that ledger will be decoupled from persistence (here, DB).  there are some useful things DB offers that we will need to generalize: 

For account balance i don't need to fetch entire ledger history. I can just do it in a single query on DB. so the persistence layer needs some more control over the getAccountBalance functionality if it supports something like this.